kong/router/atc.lua

```diff
@@ -643,37 +533,43 @@ end
 
 
 function _M:exec(ctx)
-  local req_method = get_method()
+  local fields = self.fields
+
+  local req_method = fields["http.method"] and get_method() or nil
   local req_uri = ctx and ctx.request_uri or var.request_uri
   local req_host = var.http_host
-  local sni = server_name()
+  local sni = fields["tls.sni"] and server_name() or nil
 
-  local headers, headers_key
-  if not is_empty_field(self.header_fields) then
+  local headers
+  if not is_empty_field(fields[HTTP_HEADERS_PREFIX]) then
```

Comment:
We can make this lazy, on first callback of `http.headers.foo`, call `get_http_params` to cache the header table inside params and subsequent calls to `http.headers.foo` simply uses the cached table inside params instead.

Reply:
done.